### ISSUE TITLE
Very rare occurences of NegativeArraySizeException in MapTileCache.clear

### ISSUE THREAD

**Issue by secaw** [osmdroid/osmdroid#1012]
reported in the Developer Console.
I have no idea how to reproduce this, but maybe we should just check against or catch it...

## Issue Type

[X] Bug

## Description and/or steps/code to reproduce the problem

```
java.lang.NegativeArraySizeException:  
  at org.osmdroid.util.MapTileList.<init> (SourceFile:25) 
  at org.osmdroid.tileprovider.MapTileCache.clear (SourceFile:149) 
  at org.osmdroid.tileprovider.MapTileProviderBase.clearTileCache (SourceFile:269) 
  at org.osmdroid.tileprovider.MapTileProviderBase.getMapTile (SourceFile:71)
                                                   detach
```

### If it's a bug, version(s) of android this affects:

All

### Version of osmdroid the issue relates to:

6.0.0

**Comment by monsieurtanuki** [comment]
@secaw Very strange...
We're talking about
````java
final MapTileList list = new MapTileList(mCachedTiles.size());
````
that calls
````java
public MapTileList(final int initialCapacity) {
    mTileIndices = new long[initialCapacity];
}
````
How can we have a `HashMap` (`mCachedTiles`) with a negative `size()`? Maybe it's a concurrency side effect.

As you don't know how to reproduce this issue, the fix would be something like
````java
public MapTileList(final int initialCapacity) {
    mTileIndices = new long[Math.max(0, initialCapacity)];
}
````
with a bit of `synchronized (mCachedTiles)` in `MapTileCache.clear()`

**Comment by secaw** [comment]
@monsieurtanuki Yes, it's weird. Might be related to Overlay also using `MapTileProviderBase`, but still... - shouldn't it have it's own `MapTileCache`? Solution sounds reasonable anyways.

**Comment by spyhunter99** [comment]
Definitely a concurrency issue

**Comment by monsieurtanuki** [comment]
Allegedly fixed in #1018 (could not reproduce the issue before).
@secaw As it looks like a very rare bug I presume you cannot actually test the fix. Therefore I close the issue; feel free to reopen it.

**Comment by secaw** [comment]
@monsieurtanuki Great, thanks!

**Comment by lfielke** [comment]
I've noticed instances of this crash even after the pull request merged in 5264f19ff2b5c36c0899c1d2c23f796b4785caba

There is still one method that uses `mCachedTiles` without a synchronized block: `org.osmdroid.tileprovider.MapTileCache#remove(long)`

A multithreaded unit test that calls `mapTileCache.remove(j)` with sequential values can reliably replicate the issue. It looks like the `remove` calls race, and sometimes end up double-decrementing the HashMap's size field.

```java
    @Test
    public void testConcurrency() throws InterruptedException {
        final MapTileCache mapTileCache = new MapTileCache(100);
        List<Thread> threads = new ArrayList<>();
        final Drawable dummyDrawable = new ColorDrawable(0x0);

        final int NUM_TILES = 10000;
        for (int i = 0; i < NUM_TILES; i++) {
            mapTileCache.putTile(i, dummyDrawable);
        }

        for (int i = 0; i < 10; i++) {
            Thread thread = new Thread() {
                @Override
                public void run() {
                    for (int j = 0; j < NUM_TILES; j++) {
                        mapTileCache.remove(j);
                    }
                }
            };
            thread.start();
            threads.add(thread);
        }

        for (Thread thread : threads)
            thread.join();

        mapTileCache.clear();
    }
```

Syncronizing the `drawable = mCachedTiles.remove(pMapTileIndex);` call with `mCachedTiles` makes the test case pass. I'm not sure if that call was left out on purpose for other reasons (performance/deadlock?) or if it was simply missed.

**Comment by monsieurtanuki** [comment]
@lfielke Thank you for your comment and your code. I'll have a look at it within a week. Of course there's a big focus on performance, but we cannot let the app crash.

**Comment by monsieurtanuki** [comment]
@lfielke You're right, the idea is to edit method `MapTileCache.remove` like this:
````java
public void remove(final long pMapTileIndex) {
	final Drawable drawable;
	synchronized (mCachedTiles) { // fix
		drawable = mCachedTiles.remove(pMapTileIndex);
	} // fix
	if (getTileRemovedListener() != null)
		getTileRemovedListener().onTileRemoved(pMapTileIndex);
	BitmapPool.getInstance().asyncRecycle(drawable);
}
````
Do you want to create a PR? If you don't have time I can do it quickly.
Should we also add `synchronized` on `size()` calls?

**Comment by lfielke** [comment]
Re size, `getSize()` is only called by tests at the moment, and `garbageCollection()` shouldn't matter if it has an old size value, because I it's only at the start of the method to work out a limit of how many tiles to remove.

**Comment by monsieurtanuki** [comment]
Looks OK.

**Comment by lfielke** [comment]
I can confirm the crashes are no longer occurring since the fix referenced above (in #1041).
